lib/services/budget_service.rb

```diff
@@ -0,0 +1,17 @@
+class BudgetService
+  class << self
+    def check_balance_for_update(user, quantity, event, id)
+      point = BonusEngine::Point.find(id)
+      initial_balance = event.stats_for(user)[:balance]
+
+      (initial_balance == 0 && point.quantity >= quantity) ||
```

Comment:
this is not easily readable, you could create small methods for this with a question mark, maybe, you could assign the initial_balance to an instance attribute and use it
```

def check_balance_for_update ...
   @initial_balance = event.stats_for....
   @point - Bonusengin.....
   @quantity = quantity
    is_valid_quantity? || has_balance? || exceed_points?
end

private
def is_valid_quantity?
  @initial_balance == 0 & @point.quantity >= @quantity
end

def has_balance?
  @initial_balance > @quantity
end

...
```

just check the required values as instance values ... and update the naming to be realistic xD ....